detox/src/realms/DetoxContext.js

```diff
@@ -0,0 +1,180 @@
+     */
+    this[symbols.logger] = new DetoxLogger({
+      file: temporary.for.jsonl(`${this[$sessionState].id}.${process.pid}`),
+      userConfig: this[$sessionState].detoxConfig
+        ? this[$sessionState].detoxConfig.logger
+        : null,
+    });
+
+    this.log = this[symbols.logger].child({ cat: 'user' });
+    installLegacyTracerInterface(this.log, this);
+
+    /** @type {import('../DetoxWorker') | null} */
+    this[$worker] = null;
+  }
+
+  //#region Public members
+  device = funpermaproxy(() => this[symbols.worker].device);
+
+  element = funpermaproxy.callable(() => this[symbols.worker].element);
+
+  waitFor = funpermaproxy.callable(() => this[symbols.worker].waitFor);
+
+  expect = funpermaproxy.callable(() => this[symbols.worker].expect);
+
+  by = funpermaproxy(() => this[symbols.worker].by);
+
+  web = funpermaproxy.callable(() => this[symbols.worker].web);
+
+  get DetoxConstants() {
+    return DetoxConstants;
+  }
+
+  //#endregion
+
+  //#region Internal members
+  [symbols.onRunDescribeStart] = (...args) => this[symbols.worker].onRunDescribeStart(...args);
+  [symbols.onTestStart] = (...args) => this[symbols.worker].onTestStart(...args);
+  [symbols.onHookFailure] = (...args) => this[symbols.worker].onHookFailure(...args);
+  [symbols.onTestFnFailure] = (...args) => this[symbols.worker].onTestFnFailure(...args);
+  [symbols.onTestDone] = (...args) => this[symbols.worker].onTestDone(...args);
+  [symbols.onRunDescribeFinish] = (...args) => this[symbols.worker].onRunDescribeFinish(...args);
+  [symbols.config] = funpermaproxy(() => this[symbols.session].detoxConfig);
+  [symbols.session] = funpermaproxy(() => this[$sessionState]);
+  [symbols.tracing] = Object.freeze({
+    createEventStream: () => {
+      const streamUtils = require('../logger/utils/streamUtils');
+      return streamUtils
+        .uniteSessionLogs(this[$sessionState].id)
+        .pipe(streamUtils.chromeTraceStream());
+    },
+  });
+  /** @abstract */
+  [symbols.reportTestResults](_testResults) {}
+  /**
+   * @abstract
+   * @param {Partial<DetoxInternals.DetoxInitOptions>} _opts
+   * @returns {Promise<DetoxInternals.RuntimeConfig>}
+   */
+  async [symbols.resolveConfig](_opts) { return null; }
+
+  [symbols.getStatus]() {
+    return this[$status];
+  }
+
+  get [symbols.worker]() {
+    if (!this[$worker]) {
+      throw new DetoxRuntimeError({
+        message: `Detox worker instance has not been installed in this context (${this.constructor.name}).`,
+        hint: DetoxRuntimeError.reportIssueIfJest + '\n' + 'Otherwise, make sure you call detox.installWorker() beforehand.',
+      });
+    }
+
+    return this[$worker];
+  }
+
+  /**
+   * @abstract
+   * @param {Partial<DetoxInternals.DetoxInitOptions>} [_opts]
+   * @returns {Promise<void>}
+   */
+  async [symbols.init](_opts = {}) {}
+
+  /**
+   * @param {Partial<DetoxInternals.DetoxInstallWorkerOptions>} [opts]
+   */
+  async [symbols.installWorker](opts) {
```

Comment:
Maybe it could work, indeed, if I understood you correctly, i.e., creating an artificial secondary context under the same roof as the primary one.

To my taste, it also creates another, just a different mental load. Maybe a smaller one, but still.

I prefer to think of shallow/complete Detox contexts – it makes it easier to grasp.

As for the entire routine of installing/uninstalling workers, maybe it will undergo some changes in the next versions, who knows – e.g. when we approach the capability to run multiple devices + multiple apps in the same test.

My motivation was also to keep Detox init super simple so that you can create a plain JS file and use Detox **without a test runner** at all:

```js
import detox from 'detox/internals';

async function main() {
  try {
    await detox.init();
    await device.launchApp();
    await element(by.id('someBtn')).tap();
  } finally {
    await detox.cleanup();
  }
}
```